### PR TITLE
[lang] Add tests for refine_coordinates

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #2327

A follow-up of #2327. Add tests for the JIT generated `refine_coordinates()`. Hopefully this servers as a better *documentation* for how the list-gen process works.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/ir/ir_builder.h [hunk 1]
```diff
@@ -92,8 +92,7 @@ class IRBuilder {
   template <typename XStmt>
   [[nodiscard]] LoopGuard get_loop_guard(XStmt *loop) {
     return LoopGuard(*this, loop);
-  }
-  [[nodiscard]] IfGuard get_if_guard(IfStmt *if_stmt, bool true_branch) {
+  }[[nodiscard]] IfGuard get_if_guard(IfStmt *if_stmt, bool true_branch) {
```
Comment (xumingkuan) [inline taichi/ir/ir_builder.h:95]: Revert this?

File: tests/cpp/codegen/refine_coordinates_test.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    auto sc = std::make_unique<StructCompilerLLVM>(arch_, &config_, tlctx_);
+    sc->run(*root_snode_, /*host=*/true);
+  }
+
+  Arch arch_;
+  CompileConfig config_;
+  // We shouldn't need a Program instance in this test. Unfortunately, a few
+  // places depend on the global |current_program|, so we have to.
+  // ¯\_(ツ)_/¯
```
Comment (xumingkuan) [inline tests/cpp/codegen/refine_coordinates_test.cpp:126]: ¯\\\_(ツ)\_/¯

**Review by xumingkuan — APPROVED** [review]
LGTM!

File: taichi/ir/ir_builder.h (hunk 1 above)
Comment (k-ye) [reply taichi/ir/ir_builder.h:95]: Done

**Pull request merged**